content/docs/rfcs/75/README.md

```diff
@@ -0,0 +1,173 @@
+
+# Background / Rationale / Motivation
+
+Nodes who want to synchronize with each other should be able to do so with minimal amount of communication and bandwidth usage. Existing methods present in Waku Store protocol are based on full message delivery using time range query where there is no way to know apart from a given timestamp range if other messages are also missing, making them unreliable and bandwidth consuming. Node that wants to synchronize sends a request to the peer node, and in return peer node send the response to the requesting node. After a brief low-bandwidth consuming back and flow exchange, resulting will be a list values (message_hashes) that the requesting node needs to sync with the peer node. These hashes can be used by other parts of the Store protocol to make sure the related full messages are wired to the requesting node and stored in the Store, making the requesting node fully synchronized with the peer node.
+
+A detailed view in the process can be tracked using the research [issue](https://github.com/waku-org/research/issues/62).
+
+# Theory / Semantics
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “NOT RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+Consider a request-response protocol with two roles: a client and a server.
+Client and Server both MUST be peers supporting the Waku Sync protocol. There is no other eligibility for the node at this point in time to be a client or a server who can synchronize with each other.
+
+# Wire Format Specification / Syntax
+
+```
+
+    syntax = "proto3";
+
+    package prollytree;
+
+    message Node {
+        string data = 1;              // The data stored in the node
+        int64 timestamp = 2;          // Timestamp or other unique identifier
+        string node_hash = 3;         // Hash of the node
+        int32 level = 4;              // Level of the node in the tree
+        string merkel_hash = 5;       // Merkel hash of the node
+        bool is_tail = 6;             // Flag indicating if it's a tail node
+
+        // Relationships (IDs or indices of related nodes)
+        // Note: Relationships like left, right, up, and down are typically represented using references
+        // (like indices or IDs) rather than directly embedding the objects to avoid deep nesting.
+        // In our case we can use the key of the node itself and figure out the level based on the current node level.
+        int32 left = 7;               // Reference to the left node
+        int32 right = 8;              // Reference to the right node
+        int32 up = 9;                 // Reference to the parent node
+        int32 down = 10;              // Reference to the child node
+    }
+
+    // Request message for GetRoot RPC. No fields needed for this example.
+    message GetRootRequest {
```

Comment:
> this will give flexibility for scenarios which don't depend on any store nodes.

I am not sure I am following this part? could you please elaborate more on this point?

Just to provide more context, a client sends a request to a server (here client and server are peer nodes) to get its root so, here the requesting node (client) holds the responsibility of computation based on root received. For server it should be a passive process. With `ExchangeRootRequest` I looks like client is also sending it's root, which should not be the case.

Reply:
> so that the other knows who is ahead, and decides which one should be client and which one should be server.

so basically, when an operator goes offline, they somehow know to have missed messages while they were offline, in this case the operator sends a query request based on the last available timestamp and gets messages in the old traditional way. Sync is highly useful in the case where the connection is ad-hoc that an operator might miss some messages if not all. This type of Sync request/root-check should be done in a cronjob type within certain intervals.

> it will make community based sync possible and remove the dependency for store node.

This seems like a multidimensional question, you are looking at it from Status/chat perspective, it should be dealt in app level policies in Waku client of how and where to create the Prolly tree.

Reply:
I think it's possible to save bandwidth by having the first message contain the requester's own root node. This means that initially the protocol doesn't quite behave as client-server in the first message. Each node can use this to periodically request the roots of peers to determine if it has any gaps. The responder can use this first message to also determine if it has any gaps. Both sides then have the option to initiate the sync process, but this time as client of the interaction until it has retrieved the hashes of missed messages.